### PR TITLE
Add general support for LTTng tracing and traces for EFA provider

### PR DESCRIPTION
**Pull request by YonatanNachum**
Add support for LTTng: The LTTng User Space Tracing (LTTng-UST) library allows any C/C++ application to be instrumented for and traced by LTTng.

Add CMake flag to compile in LTTng support for providers who added LTTng traces, without suppling this flags the tracing code will be compiled out without any influence on other code.

Also add initial support for EFA tracing: post receive, send and completion.

### CODE REVIEW

File: providers/efa/CMakeLists.txt [hunk 1]
```diff
@@ -9,3 +14,8 @@ publish_headers(infiniband
 )
 
 rdma_pkg_config("efa" "libibverbs" "${CMAKE_THREAD_LIBS_INIT}")
+
+if (ENABLE_TRACE AND LTTNGUST_FOUND)
+	target_include_directories(efa PUBLIC ".")
```
Comment (jgunthorpe) [inline providers/efa/CMakeLists.txt:19]: Huh?

File: providers/efa/efa_trace.c [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* SPDX-License-Identifier: GPL-2.0 OR BSD-2-Clause */
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#define LTTNG_UST_TRACEPOINT_CREATE_PROBES
+#define LTTNG_UST_TRACEPOINT_DEFINE
+
+#include "./efa_trace.h"
```
Comment (jgunthorpe) [inline providers/efa/efa_trace.c:9]: No need for ./

File: CMakeLists.txt [hunk 3]
```diff
@@ -556,6 +558,21 @@ if (DRM_INCLUDE_DIRS)
   endif()
 endif()
 
+# Tracer support
+if (DEFINED ENABLE_TRACE)
+  message(STATUS "ENABLE_TRACE is set: ${ENABLE_TRACE}")
+  if ("${ENABLE_TRACE}" STREQUAL "lttng")
+    include(FindLTTngUST REQUIRED)
+    if (NOT LTTNGUST_FOUND)
```
Comment (jgunthorpe) [inline CMakeLists.txt:566]: REQUIRED already does this error handling, doesn't it?

File: CMakeLists.txt [hunk 4]
```diff
@@ -556,6 +558,21 @@ if (DRM_INCLUDE_DIRS)
...
+      message(FATAL_ERROR "lttng-UST libraries not found")
+    endif()
+    add_definitions(-DTRACE_ENABLED)
+    add_definitions(-DTRACE_LTTNG)
+  else()
+    message(FATAL_ERROR "Tracer isn't supported: ${ENABLE_TRACE}")
```
Comment (jgunthorpe) [inline CMakeLists.txt:572]: Why would we support more than one kind of tracer?

**Review by jgunthorpe — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 5]
```diff
@@ -556,6 +558,21 @@ if (DRM_INCLUDE_DIRS)
   endif()
 endif()
 
+# Tracer support
+if (DEFINED ENABLE_TRACE)
+  message(STATUS "ENABLE_TRACE is set: ${ENABLE_TRACE}")
```
[the remark this answers is not in the document]
Comment (YonatanNachum) [reply CMakeLists.txt]: Done

File: providers/efa/efa_trace.c (hunk 2 above)
Comment (YonatanNachum) [reply providers/efa/efa_trace.c]: Done

File: providers/efa/CMakeLists.txt (hunk 1 above)
Comment (YonatanNachum) [reply providers/efa/CMakeLists.txt:19]: This is needed for the tracepoint provider header file: efa_trace.h, LTTng needs each header file to supply the path to the header itself:
```
#define LTTNG_UST_TRACEPOINT_INCLUDE "efa_trace.h"
```
without It, I get the error:
```
fatal error: efa_trace.h: No such file or directory
 #define LTTNG_UST_TRACEPOINT_INCLUDE "efa_trace.h"
```

I tried to change to "efa/efa_trace.h" and "providers/efa/efa_trace.h" and got the same error.

If you look at LTTng-UST github example for CMake: [link](https://github.com/lttng/lttng-ust/blob/master/doc/examples/cmake-multiple-shared-libraries/CMakeLists.txt), you can see that they also used:
```
include_directories(.)
```

**Comment by YonatanNachum** [comment]
Hey @jgunthorpe, kind reminder for this patch.

Thanks.

**Comment by jgunthorpe** [comment]
I still think we should not try to support multiple tracers? Is there some reason? It isn't code transparent either, right?

**Comment by mrgolin** [comment]
> I still think we should not try to support multiple tracers? Is there some reason? It isn't code transparent either, right?

It should be but I'm ok with supporting only LTTng and removing some of the configuration logic, at least until someone will really need this.

**Comment by YonatanNachum** [comment]
Hey, @jgunthorpe, Kind reminder.

Thanks.

**Pull request merged**